File: sem iv/Theory of Computaion/lab/code/c/11_pda_0s_followed_1s_finalstate.cpp
// PDA accepting L = { w | w in {0,1}* and 0's are followed by an equal number of 1's }
// i.e. L = { 0^n 1^n }, acceptance by final state
#include <iostream>
using namespace std;
#define MAX 100

char symbolStack[MAX];
int top = -1;

void push(char ch) {
    if (top < MAX - 1)
        symbolStack[++top] = ch;
    else
        cout << "Stack is full\n";
}

char pop() {
    if (top > -1)
        return symbolStack[top--];
    cout << "Stack is empty\n";
    return '\0';
}

char getTop() {
    return symbolStack[top];
}

int main() {
    string input;
    cout << "Enter string: ";
    cin >> input;

    push('$');
    int state = 0; // 0: reading 0's, 1: reading 1's
    bool valid = true;

    for (char ch : input) {
        if (state == 0) {
            if (ch == '0') {
                push('0');
            } else if (ch == '1') {
                if (getTop() == '0') { pop(); state = 1; }
                else { valid = false; break; }
            } else { valid = false; break; }
        } else {
            if (ch == '1') {
                if (getTop() == '0') pop();
                else { valid = false; break; }
            } else { valid = false; break; }
        }
    }

    if (valid && getTop() == '$')
        cout << input << " is accepted" << endl;
    else
        cout << input << " is rejected" << endl;
    return 0;
}

File: sem iv/Theory of Computaion/lab/code/python/11_pda_0s_followed_1s_finalstate.py
# PDA accepting L = { w | w in {0,1}* and 0's are followed by an equal number of 1's }
# i.e. L = { 0^n 1^n }, acceptance by final state

def main():
    inp = input("Enter string: ")
    stack = ['$']
    state = 0  # 0: reading 0's, 1: reading 1's
    valid = True

    for ch in inp:
        if state == 0:
            if ch == '0':
                stack.append('0')
            elif ch == '1':
                if stack[-1] == '0':
                    stack.pop()
                    state = 1
                else:
                    valid = False
                    break
            else:
                valid = False
                break
        else:
            if ch == '1':
                if stack[-1] == '0':
                    stack.pop()
                else:
                    valid = False
                    break
            else:
                valid = False
                break

    if valid and stack[-1] == '$':
        print(f"{inp} is accepted")
    else:
        print(f"{inp} is rejected")

if __name__ == "__main__":
    main()
